File: long_questions/DSA/Leetcode/golang/14_Backtracking/046_Permutations.go
package main

import "fmt"

// 46. Permutations
// Time: O(N * N!), Space: O(N!) for result + O(N) for recursion
func permute(nums []int) [][]int {
	var result [][]int
	used := make([]bool, len(nums))
	current := make([]int, len(nums))
	
	backtrackPermute(nums, used, current, 0, &result)
	return result
}

func backtrackPermute(nums []int, used []bool, current []int, pos int, result *[][]int) {
	if pos == len(nums) {
		// Make a copy of current permutation
		temp := make([]int, len(nums))
		copy(temp, current)
		*result = append(*result, temp)
		return
	}
	
	for i := 0; i < len(nums); i++ {
		if !used[i] {
			used[i] = true
			current[pos] = nums[i]
			
			backtrackPermute(nums, used, current, pos+1, result)
			
			used[i] = false
		}
	}
}

// Alternative approach using swapping
func permuteSwap(nums []int) [][]int {
	var result [][]int
	backtrackSwap(nums, 0, &result)
	return result
}

func backtrackSwap(nums []int, start int, result *[][]int) {
	if start == len(nums) {
		// Make a copy of current permutation
		temp := make([]int, len(nums))
		copy(temp, nums)
		*result = append(*result, temp)
		return
	}
	
	for i := start; i < len(nums); i++ {
		nums[start], nums[i] = nums[i], nums[start]
		backtrackSwap(nums, start+1, result)
		nums[start], nums[i] = nums[i], nums[start]
	}
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Backtracking for Permutations
- **Position-based Selection**: Fill each position with unused elements
- **Used Tracking**: Track which elements have been used in current permutation
- **Recursive Building**: Build permutations position by position
- **Backtracking**: Remove choices when exploring different paths

## 2. PROBLEM CHARACTERISTICS
- **Permutation Generation**: Generate all possible orderings of elements
- **Unique Elements**: Input contains distinct elements
- **Complete Permutations**: Each permutation uses all elements exactly once
- **Order Matters**: Different orderings are distinct permutations

## 3. SIMILAR PROBLEMS
- Permutations II (LeetCode 47) - Handle duplicate elements
- Next Permutation (LeetCode 31) - Find next lexicographic permutation
- Permutation Sequence (LeetCode 60) - Find kth permutation
- Subsets (LeetCode 78) - Generate all subsets

## 4. KEY OBSERVATIONS
- **Backtracking natural fit**: Need to explore all possible arrangements
- **Used array essential**: Track which elements are available for current position
- **Position parameter**: Track current position being filled
- **Complete permutation**: Base case when all positions are filled

## 5. VARIATIONS & EXTENSIONS
- **Duplicate Elements**: Handle repeated elements in input
- **Partial Permutations**: Generate permutations of length k
- **Lexicographic Order**: Generate permutations in sorted order
- **Kth Permutation**: Find specific permutation without generating all

## 6. INTERVIEW INSIGHTS
- Always clarify: "Are elements unique? What about empty input?"
- Edge cases: empty array, single element, large arrays
- Time complexity: O(N × N!) - N! permutations, each takes O(N) to copy
- Space complexity: O(N) for recursion + O(N × N!) for result

## 7. COMMON MISTAKES
- Not using used array (causes duplicates in permutations)
- Making shallow copies of current permutation
- Not resetting used array properly during backtracking
- Using wrong base case condition
- Not handling empty input case

## 8. OPTIMIZATION STRATEGIES
- **Swap method**: More memory efficient, no extra used array
- **In-place modification**: Modify array directly instead of using extra space
- **Early pruning**: Not applicable for complete permutations
- **Iterative generation**: Use Heap's algorithm for iterative approach

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like arranging books on a shelf:**
- You have N distinct books (array elements)
- You want to find all possible ways to arrange them
- For each position, you choose from the remaining books
- Once you place a book, you can't use it again in current arrangement
- Continue until all positions are filled, then backtrack to try different arrangements

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Array of distinct elements
2. **Goal**: Generate all possible permutations (orderings)
3. **Output**: List of all permutations, each containing all elements
4. **Constraint**: Each element used exactly once per permutation

#### Phase 2: Key Insight Recognition
- **"Backtracking natural fit"** → Need to explore all possible arrangements
- **"Position-based selection"** → Fill each position with available elements
- **"Used tracking essential"** → Need to track which elements are used
- **"Complete coverage"** → Each permutation must use all elements

#### Phase 3: Strategy Development
```
Human thought process:
"I need to generate all permutations of the array.
I'll use backtracking to build permutations position by position.
For each position, I'll try each unused element.
I'll mark elements as used when I include them.
When I reach the end, I have a complete permutation.
Then I backtrack and try different choices."
```

#### Phase 4: Edge Case Handling
- **Empty array**: Return empty result
- **Single element**: Return [[element]]
- **Large arrays**: Handle recursion depth and memory appropriately
- **Duplicate elements**: Need special handling (Permutations II)

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Input: [1,2,3]

Human thinking:
"I'll build permutations position by position:

Position 0:
- Try 1 (unused): current=[1,_ ,_]
  Mark 1 as used
  Position 1:
  - Try 2 (unused): current=[1,2,_]
    Mark 2 as used
    Position 2:
    - Try 3 (unused): current=[1,2,3] → COMPLETE [1,2,3]
    Backtrack: unmark 3
  Backtrack: unmark 2
  - Try 3 (unused): current=[1,3,_]
    Mark 3 as used
    Position 2:
    - Try 2 (unused): current=[1,3,2] → COMPLETE [1,3,2]
    Backtrack: unmark 2
  Backtrack: unmark 3
Backtrack: unmark 1

Position 0:
- Try 2 (unused): current=[2,_ ,_]
  Mark 2 as used
  Position 1:
  - Try 1 (unused): current=[2,1,_]
    Mark 1 as used
    Position 2:
    - Try 3 (unused): current=[2,1,3] → COMPLETE [2,1,3]
  Backtrack: unmark 1
  - Try 3 (unused): current=[2,3,_]
    Mark 3 as used
    Position 2:
    - Try 1 (unused): current=[2,3,1] → COMPLETE [2,3,1]
  Backtrack: unmark 3
Backtrack: unmark 2

Position 0:
- Try 3 (unused): current=[3,_ ,_]
  Similar process generates [3,1,2] and [3,2,1]

All permutations: [1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], [3,2,1]"
```

#### Phase 6: Intuition Validation
- **Why backtracking works**: Need to explore all possible arrangements
- **Why used array works**: Prevents reusing elements in same permutation
- **Why O(N × N!) time**: N! permutations, each takes O(N) to copy
- **Why position-based works**: Natural way to build permutations

### Common Human Pitfalls & How to Avoid Them
1. **"Why not use next_permutation?"** → That finds the next one, not all permutations
2. **"Should I use recursion or iteration?"** → Recursion is more intuitive for this problem
3. **"What about very large N?"** → Complexity is inherently factorial, can't do much better
4. **"Can I optimize space?"** → Use swap method to avoid used array

### Real-World Analogy
**Like finding all possible seating arrangements:**
- You have N distinct people (array elements)
- You need to find all possible ways to seat them in N chairs
- For each chair, you choose from the remaining unseated people
- Once someone is seated, they can't be seated again in current arrangement
- Continue until all chairs are filled, then try different arrangements

### Human-Readable Pseudocode
```
function permute(nums):
    result = []
    used = [false] * length(nums)
    current = [0] * length(nums)
    
    backtrack(nums, used, current, 0, result)
    return result

function backtrack(nums, used, current, pos, result):
    if pos == length(nums):
        add copy of current to result
        return
    
    for i from 0 to length(nums)-1:
        if not used[i]:
            used[i] = true
            current[pos] = nums[i]
            
            backtrack(nums, used, current, pos + 1, result)
            
            used[i] = false  // Backtrack
```

### Execution Visualization

### Example: nums = [1,2,3]
```
Recursion Tree (Used Array Method):
                    [], pos=0
                   /    |    \
                [1]    [2]    [3]
                pos=1   pos=1   pos=1
               / |     / |     / |
            [1,2] [1,3] [2,1] [2,3] [3,1] [3,2]
            pos=2   pos=2   pos=2   pos=2   pos=2   pos=2
             |      |      |      |      |      |
           [1,2,3] [1,3,2] [2,1,3] [2,3,1] [3,1,2] [3,2,1]

All permutations found at pos=3 (base case)
```

### Key Visualization Points:
- **Position-based building**: Each level fills one position
- **Used tracking**: Elements marked as used when included
- **Backtracking**: Elements unmarked when exploring different paths
- **Complete permutations**: Base case when all positions filled
- **Systematic exploration**: Try all unused elements at each position

### Memory Layout Visualization:
```
Used Array Evolution:
[false,false,false] → [true,false,false] → [true,true,false] → [true,true,true]
                      ↘ [true,false,true] → [true,false,true]
                      ↘ [false,true,false] → [false,true,true]
                      ↘ [false,false,true] → [false,true,true]

Current Array Evolution:
[_,_,_] → [1,_,_] → [1,2,_] → [1,2,3] ✓
         → [1,_,_] → [1,3,_] → [1,3,2] ✓
         → [2,_,_] → [2,1,_] → [2,1,3] ✓
         → [2,_,_] → [2,3,_] → [2,3,1] ✓
         → [3,_,_] → [3,1,_] → [3,1,2] ✓
         → [3,_,_] → [3,2,_] → [3,2,1] ✓
```

### Time Complexity Breakdown:
- **Number of permutations**: N! (factorial)
- **Copy time per permutation**: O(N) to copy current permutation
- **Total time**: O(N × N!)
- **Space complexity**: O(N) for recursion stack + O(N × N!) for result
- **Used array**: O(N) additional space

### Alternative Approaches:

#### 1. Swap Method (O(N × N!) time, O(N) space)
```go
func permuteSwap(nums []int) [][]int {
    var result [][]int
    backtrackSwap(nums, 0, &result)
    return result
}

func backtrackSwap(nums []int, start int, result *[][]int) {
    if start == len(nums) {
        // Make a copy of current permutation
        temp := make([]int, len(nums))
        copy(temp, nums)
        *result = append(*result, temp)
        return
    }
    
    for i := start; i < len(nums); i++ {
        nums[start], nums[i] = nums[i], nums[start]
        backtrackSwap(nums, start+1, result)
        nums[start], nums[i] = nums[i], nums[start]
    }
}
```
- **Pros**: No extra used array, more memory efficient
- **Cons**: Modifies input array, need careful swapping

#### 2. Iterative Heap's Algorithm (O(N × N!) time, O(N) space)
```go
func permuteHeap(nums []int) [][]int {
    var result [][]int
    n := len(nums)
    c := make([]int, n) // Control array
    
    // Add initial permutation
    temp := make([]int, n)
    copy(temp, nums)
    result = append(result, temp)
    
    i := 0
    for i < n {
        if c[i] < i {
            if i%2 == 0 {
                nums[0], nums[i] = nums[i], nums[0]
            } else {
                nums[c[i]], nums[i] = nums[i], nums[c[i]]
            }
            
            temp := make([]int, n)
            copy(temp, nums)
            result = append(result, temp)
            
            c[i]++
            i = 0
        } else {
            c[i] = 0
            i++
        }
    }
    
    return result
}
```
- **Pros**: Iterative, no recursion stack
- **Cons**: Complex to understand and implement

#### 3. Lexicographic Generation (O(N × N!) time, O(1) extra space)
```go
func permuteLexicographic(nums []int) [][]int {
    sort.Ints(nums)
    var result [][]int
    
    // Add initial permutation
    temp := make([]int, len(nums))
    copy(temp, nums)
    result = append(result, temp)
    
    for nextPermutation(nums) {
        temp := make([]int, len(nums))
        copy(temp, nums)
        result = append(result, temp)
    }
    
    return result
}

func nextPermutation(nums []int) bool {
    n := len(nums)
    
    // Find the first element from the right that is smaller than the next element
    i := n - 2
    for i >= 0 && nums[i] >= nums[i+1] {
        i--
    }
    
    if i < 0 {
        return false // Already the last permutation
    }
    
    // Find the smallest element greater than nums[i] from the right
    j := n - 1
    for nums[j] <= nums[i] {
        j--
    }
    
    // Swap them
    nums[i], nums[j] = nums[j], nums[i]
    
    // Reverse the suffix
    reverse(nums, i+1, n-1)
    
    return true
}

func reverse(nums []int, left, right int) {
    for left < right {
        nums[left], nums[right] = nums[right], nums[left]
        left++
        right--
    }
}
```
- **Pros**: Generates permutations in lexicographic order
- **Cons**: More complex implementation, requires sorting first

### Extensions for Interviews:
- **Permutations II**: Handle duplicate elements in input
- **Next Permutation**: Find next lexicographic permutation
- **Kth Permutation**: Find kth permutation without generating all
- **Partial Permutations**: Generate permutations of length k
- **Circular Permutations**: Generate circular arrangements
*/
func main() {
	// Test cases
	testCases := [][]int{
		{1, 2, 3},
		{0, 1},
		{1},
		{},
		{1, 2, 3, 4},
		{1, 1, 2},
		{1, 2, 2},
		{5, 6, 7},
		{1, 2, 3, 4, 5},
	}
	
	for i, nums := range testCases {
		// Make copies for both methods
		nums1 := make([]int, len(nums))
		copy(nums1, nums)
		nums2 := make([]int, len(nums))
		copy(nums2, nums)
		
		result1 := permute(nums1)
		result2 := permuteSwap(nums2)
		
		fmt.Printf("Test Case %d: %v\n", i+1, nums)
		fmt.Printf("  Used array: %d permutations\n", len(result1))
		fmt.Printf("  Swap method: %d permutations\n\n", len(result2))
	}
}

File: long_questions/DSA/Leetcode/java/14_Backtracking/046_Permutations.java
import java.util.*;

public class Permutations {
    
    // 46. Permutations
    // Time: O(N * N!), Space: O(N!) for result + O(N) for recursion
    public static List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        int[] current = new int[nums.length];
        
        backtrackPermute(nums, used, current, 0, result);
        return result;
    }

    private static void backtrackPermute(int[] nums, boolean[] used, int[] current, 
                                   int pos, List<List<Integer>> result) {
        if (pos == nums.length) {
            // Make a copy of current permutation
            List<Integer> temp = new ArrayList<>();
            for (int num : current) {
                temp.add(num);
            }
            result.add(temp);
            return;
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (!used[i]) {
                used[i] = true;
                current[pos] = nums[i];
                
                backtrackPermute(nums, used, current, pos + 1, result);
                
                used[i] = false;
            }
        }
    }

    // Alternative approach using swapping
    public static List<List<Integer>> permuteSwap(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrackSwap(nums, 0, result);
        return result;
    }

    private static void backtrackSwap(int[] nums, int start, List<List<Integer>> result) {
        if (start == nums.length) {
            // Make a copy of current permutation
            List<Integer> temp = new ArrayList<>();
            for (int num : nums) {
                temp.add(num);
            }
            result.add(temp);
            return;
        }
        
        for (int i = start; i < nums.length; i++) {
            swap(nums, start, i);
            backtrackSwap(nums, start + 1, result);
            swap(nums, start, i);
        }
    }

    private static void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    public static void main(String[] args) {
        // Test cases
        int[][] testCases = {
            {1, 2, 3},
            {0, 1},
            {1},
            {},
            {1, 2, 3, 4},
            {1, 1, 2},
            {1, 2, 2},
            {5, 6, 7},
            {1, 2, 3, 4, 5}
        };
        
        for (int i = 0; i < testCases.length; i++) {
            // Make copies for both methods
            int[] nums1 = testCases[i].clone();
            int[] nums2 = testCases[i].clone();
            
            List<List<Integer>> result1 = permute(nums1);
            List<List<Integer>> result2 = permuteSwap(nums2);
            
            System.out.printf("Test Case %d: %s\n", i + 1, Arrays.toString(testCases[i]));
            System.out.printf("  Used array: %d permutations\n", result1.size());
            System.out.printf("  Swap method: %d permutations\n\n", result2.size());
        }
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Backtracking
- **Permutation Generation**: Generate all possible arrangements
- **Recursive Exploration**: Try each unused element at each position
- **Backtracking**: Undo choices when exploring dead ends
- **Complete Solutions**: When all positions are filled

## 2. PROBLEM CHARACTERISTICS
- **Permutations**: All possible orderings of array elements
- **No Duplicates**: Each element appears exactly once
- **Complete Search**: Must explore all possibilities
- **Result Collection**: Store all valid permutations

## 3. SIMILAR PROBLEMS
- Combinations
- Subsets
- Letter Case Permutation
- N-Queens Problem

## 4. KEY OBSERVATIONS
- Number of permutations = N! for N elements
- Backtracking explores each position systematically
- Used array tracks which elements are placed
- Each recursive call fixes one position
- Time complexity is O(N!) by necessity

## 5. VARIATIONS & EXTENSIONS
- Permutations with duplicates allowed
- K-permutations (select K from N)
- Next permutation (lexicographic order)
- Permutation with constraints

## 6. INTERVIEW INSIGHTS
- Clarify: "Should I handle duplicate elements?"
- Edge cases: empty array, single element
- Time complexity: O(N!) vs O(N^N) naive
- Space complexity: O(N!) for result storage

## 7. COMMON MISTAKES
- Not handling used array correctly
- Forgetting to backtrack (undo choices)
- Creating result copies incorrectly
- Stack overflow for large N

## 8. OPTIMIZATION STRATEGIES
- Swapping approach reduces memory usage
- Iterative generation with next_permutation
- Early pruning for constrained permutations
- In-place modifications save space

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like arranging books on a shelf:**
- You have N different books (array elements)
- You want to try all possible arrangements
- For each position on shelf, choose an unused book
- If arrangement works, record it
- Backtrack and try different book for that position
- Continue until all positions are filled

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Array of distinct integers
2. **Goal**: Generate all possible permutations
3. **Output**: List of all permutations

#### Phase 2: Key Insight Recognition
- **"How many permutations?"** → N! for N elements
- **"How to generate systematically?"** → Backtracking recursion
- **"What to track?"** → Used elements and current position
- **"When to stop?"** → When all positions filled

#### Phase 3: Strategy Development
```
Human thought process:
"I'll use backtracking:
1. For each position from 0 to N-1:
   - Try each unused element at this position
   - Mark element as used
   - Recursively fill next positions
   - Backtrack: unmark element when returning
2. When all positions filled, record permutation
3. This explores all N! possibilities"
```

#### Phase 4: Edge Case Handling
- **Empty array**: Return empty list
- **Single element**: Return [[element]]
- **Large N**: May cause stack overflow
- **Memory constraints**: N! grows very quickly

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Input: [1, 2, 3]

Human thinking:
"Start with empty permutation:
position = 0, used = [false, false, false], current = []

Position 0:
- Try 1: used[0] = false, place 1
  → current = [1], used = [true, false, false]
  → Recurse to position 1

Position 1:
- Try 2: used[1] = false, place 2
  → current = [1, 2], used = [true, true, false]
  → Recurse to position 2

Position 2:
- Try 3: used[2] = false, place 3
  → current = [1, 2, 3], used = [true, true, true]
  → All positions filled, record [1, 2, 3]
  → Backtrack: remove 3, used[2] = false

Position 1 (backtrack):
- Try 3: used[2] = false, place 3
  → current = [1, 3], used = [true, false, true]
  → Recurse to position 2

Position 2:
- Try 2: used[1] = true, skip
- Try remaining options...
Continue exploring all possibilities...

Final result: [[1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], [3,2,1]] ✓"
```

#### Phase 6: Intuition Validation
- **Why it works**: Systematic exploration of all possibilities
- **Why it's complete**: Each position tries every unused element
- **Why it's correct**: Backtracking ensures no duplicates

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just generate randomly?"** → Won't get all permutations
2. **"What about iterative approach?"** → Backtracking is more natural
3. **"How to handle duplicates?"** → Need counting and skipping logic
4. **"What about memory?"** → N! grows very fast

### Real-World Analogy
**Like creating all possible schedules:**
- You have N tasks (array elements)
- You want to try all possible orderings
- For each time slot, choose an unscheduled task
- Schedule it and mark as used
- Continue until all time slots are filled
- Backtrack when a schedule doesn't work
- Each complete schedule is one permutation

### Human-Readable Pseudocode
```
function permute(nums):
    result = []
    used = [false] * nums.length
    current = [] * nums.length
    
    backtrack(nums, used, current, 0, result)
    return result

function backtrack(nums, used, current, pos, result):
    if pos == nums.length:
        result.add(copy(current))
        return
    
    for i from 0 to nums.length-1:
        if not used[i]:
            used[i] = true
            current[pos] = nums[i]
            backtrack(nums, used, current, pos + 1, result)
            used[i] = false  // backtrack
```

### Execution Visualization

### Example: nums = [1, 2, 3]
```
Backtracking Tree:
                    pos=0: try 1
                   /         pos=1: try 2
                  /           pos=2: try 3
                 /             [1,2,3] ✓
                pos=2: try 2 (used)
               pos=2: try 1 (used)
              
              pos=1: try 3
             /         pos=2: try 2
            /          [1,3,2] ✓
           pos=2: try 1
          /           [1,3,1] ✓
         
        pos=0: try 2
       /         pos=1: try 1
      /          pos=2: try 3
     /           [2,1,3] ✓
    pos=1: try 3
   /         pos=2: try 2
  /          [2,3,1] ✓
 pos=2: try 1
/           [2,3,2] ✓

All 6 permutations generated ✓
```

### Key Visualization Points:
- **Backtracking tree** explores all choice combinations
- **Used array** prevents duplicate selections
- **Position tracking** builds permutations incrementally
- **Complete solutions** when reaching leaf nodes

### Memory Layout Visualization:
```
Recursive Stack for [1,2,3]:
Level 0: permute([1,2,3], pos=0)
├─ Try 1: permute([_,2,3], pos=1)
│  ├─ Try 2: permute([1,_,3], pos=2)
│  │  └─ Try 3: [1,2,3] ✓
│  └─ Try 3: permute([1,_,2], pos=2)
│     └─ Try 2: [1,3,2] ✓
└─ Try 2: permute([1,_,3], pos=1)
   ├─ Try 1: permute([2,_,3], pos=2)
   │  └─ Try 3: [2,1,3] ✓
   └─ Try 3: permute([2,_,1], pos=2)
      └─ Try 1: [2,3,1] ✓

Used Array Evolution:
[1,2,3] → [true,true,true] → [true,true,true] → backtrack...
```

### Time Complexity Breakdown:
- **Each Level**: N choices, then N-1, then N-2, etc.
- **Total Permutations**: N! possibilities
- **Time**: O(N!) for generating all permutations
- **Space**: O(N!) for result storage + O(N) for recursion
- **Optimal**: Cannot do better than O(N!) for this problem
*/
